compiler/hash-ir/src/ir.rs

```diff
@@ -155,33 +160,77 @@ pub enum AddressMode {
     Smart,
 }
 
-/// A defined statement within the IR
 #[derive(Debug, PartialEq, Eq)]
-pub enum StatementKind<'i> {
-    /// Filler kind when expressions are optimised out or removed for other
-    /// reasons.
-    Nop,
+pub enum PlaceProjection {
+    /// When we want to narrow down the union type to some specific
+    /// variant.
+    Downcast(usize),
+    /// A reference to a specific field within the place, at this stage they
+    /// are represented as indexes into the field store of the place type.
+    Field(usize),
+    /// Take the index of some specific place, the index does not need to be
+    /// constant
+    Index(Local),
+    /// We want to dereference the place
+    Deref,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Place {
+    /// The original place of where this is referring to.
+    pub local: Local,
+    /// Any projections that are applied onto the `local` in
+    /// order to specify an exact location within the local.
+    pub projections: Vec<PlaceProjection>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum AggregateKind {
+    Tuple,
+    Array(TermId),
+    Enum(NominalDefId, usize),
+    Struct(NominalDefId),
+}
```

Comment:
So type information is available in the IR?

Reply:
Yeah it will need to be available in order to query information about fields of types, layout, etc.